File: ltsapi/api_struct.py
# -*- coding=utf-8 -*-
"""
生成结构化的代码

"""


class Base(object):
    def to_dict(self):
        """

        :return:
        """

        return {key: value for key, value in self.__dict__.iteritems() if value}

    def __repr__(self):
        """
        :return:
        """
        return "<%s>" % ",".join(["%s:%s" % (item, value) for item, value in self.__dict__.iteritems()])


class CSecurityFtdcRspInfoField(Base):
    """

    """

    def __init__(self, ErrorID, ErrorMsg):
        """

        :param ErrorID:
        :param ErrorMsg:
        :return:
        """
        self.ErrorID = int(ErrorID)
        self.ErrorMsg = ErrorMsg


class CSecurityFtdcRspUserLoginField(Base):
    """

    """

    def __init__(self, TradingDay, LoginTime, BrokerID, UserID, SystemName, FrontID, SessionID, MaxOrderRef):
        """

        :param TradingDay:
        :param LoginTime:
        :param BrokerID:
        :param UserID:
        :param SystemName:
        :param FrontID:
        :param SessionID:
        :param MaxOrderRef:
        :return:
        """
        self.TradingDay = TradingDay
        self.LoginTime = LoginTime
        self.BrokerID = BrokerID
        self.UserID = UserID
        self.SystemName = SystemName
        self.FrontID = int(FrontID)
        self.SessionID = int(SessionID)
        self.MaxOrderRef = MaxOrderRef


class CSecurityFtdcUserLogoutField(Base):
    """

    """

    def __init__(self, BrokerID, UserID):
        """

        :param BrokerID:
        :param UserID:
        :return:
        """
        self.BrokerID = BrokerID
        self.UserID = UserID


class CSecurityFtdcSpecificInstrumentField(Base):
    def __init__(self, InstrumentID, ExchangeID):
        """

        :param InstrumentID:
        :param ExchangeID:
        :return:
        """
        self.InstrumentID = InstrumentID
        self.ExchangeID = ExchangeID


class CSecurityFtdcDepthMarketDataField(Base):
    """

    """

    def __init__(self, TradingDay, InstrumentID, ExchangeID, ExchangeInstID, LastPrice, PreSettlementPrice,
                 PreClosePrice, PreOpenInterest, OpenPrice, HighestPrice, LowestPrice, Volume, Turnover, OpenInterest,
                 ClosePrice, SettlementPrice, UpperLimitPrice, LowerLimitPrice, PreDelta, CurrDelta, UpdateTime,
                 UpdateMillisec, BidPrice1, BidVolume1, AskPrice1, AskVolume1, BidPrice2, BidVolume2, AskPrice2,
                 AskVolume2, BidPrice3, BidVolume3, AskPrice3, AskVolume3, BidPrice4, BidVolume4, AskPrice4, AskVolume4,
                 BidPrice5, BidVolume5, AskPrice5, AskVolume5, AveragePrice, ActionDay):
        """

        :param TradingDay:
        :param InstrumentID:
        :param ExchangeID:
        :param ExchangeInstID:
        :param LastPrice:
        :param PreSettlementPrice:
        :param PreClosePrice:
        :param PreOpenInterest:
        :param OpenPrice:
        :param HighestPrice:
        :param LowestPrice:
        :param Volume:
        :param Turnover:
        :param OpenInterest:
        :param ClosePrice:
        :param SettlementPrice:
        :param UpperLimitPrice:
        :param LowerLimitPrice:
        :param PreDelta:
        :param CurrDelta:
        :param UpdateTime:
        :param UpdateMillisec:
        :param BidPrice1:
        :param BidVolume1:
        :param AskPrice1:
        :param AskVolume1:
        :param BidPrice2:
        :param BidVolume2:
        :param AskPrice2:
        :param AskVolume2:
        :param BidPrice3:
        :param BidVolume3:
        :param AskPrice3:
        :param AskVolume3:
        :param BidPrice4:
        :param BidVolume4:
        :param AskPrice4:
        :param AskVolume4:
        :param BidPrice5:
        :param BidVolume5:
        :param AskPrice5:
        :param AskVolume5:
        :param AveragePrice:
        :param ActionDay:
        :return:
        """
        self.TradingDay = TradingDay
        self.InstrumentID = InstrumentID
        self.ExchangeID = ExchangeID
        self.ExchangeInstID = ExchangeInstID
        self.LastPrice = float(LastPrice)
        self.PreSettlementPrice = float(PreSettlementPrice)
        self.PreClosePrice = float(PreClosePrice)
        self.PreOpenInterest = float(PreOpenInterest)
        self.OpenPrice = float(OpenPrice)
        self.HighestPrice = float(HighestPrice)
        self.LowestPrice = float(LowestPrice)
        self.Volume = int(Volume)
        self.Turnover = float(Turnover)
        self.OpenInterest = float(OpenInterest)
        self.ClosePrice = float(ClosePrice)
        self.SettlementPrice = float(SettlementPrice)
        self.UpperLimitPrice = float(UpperLimitPrice)
        self.LowerLimitPrice = float(LowerLimitPrice)
        self.PreDelta = float(PreDelta)
        self.CurrDelta = float(CurrDelta)
        self.UpdateTime = UpdateTime
        self.UpdateMillisec = int(UpdateMillisec)
        self.BidPrice1 = float(BidPrice1)
        self.BidVolume1 = int(BidVolume1)
        self.AskPrice1 = float(AskPrice1)
        self.AskVolume1 = int(AskVolume1)
        self.BidPrice2 = float(BidPrice2)
        self.BidVolume2 = int(BidVolume2)
        self.AskPrice2 = float(AskPrice2)
        self.AskVolume2 = int(AskVolume2)
        self.BidPrice3 = float(BidPrice3)
        self.BidVolume3 = int(BidVolume3)
        self.AskPrice3 = float(AskPrice3)
        self.AskVolume3 = int(AskVolume3)
        self.BidPrice4 = float(BidPrice4)
        self.BidVolume4 = int(BidVolume4)
        self.AskPrice4 = float(AskPrice4)
        self.AskVolume4 = int(AskVolume4)
        self.BidPrice5 = float(BidPrice5)
        self.BidVolume5 = int(BidVolume5)
        self.AskPrice5 = float(AskPrice5)
        self.AskVolume5 = int(AskVolume5)
        self.AveragePrice = float(AveragePrice)
        self.ActionDay = ActionDay


class CSecurityFtdcReqUserLoginField(Base):
    """

    """

    def __init__(self, TradingDay, BrokerID, UserID, Password, UserProductInfo, InterfaceProductInfo, ProtocolInfo,
                 MacAddress, OneTimePassword, ClientIPAddress, AuthCode):
        """

        :param TradingDay:
        :param BrokerID:
        :param UserID:
        :param Password:
        :param UserProductInfo:
        :param InterfaceProductInfo:
        :param ProtocolInfo:
        :param MacAddress:
        :param OneTimePassword:
        :param ClientIPAddress:
        :param AuthCode:
        :return:
        """
        self.TradingDay = TradingDay
        self.BrokerID = BrokerID
        self.UserID = UserID
        self.Password = Password
        self.UserProductInfo = UserProductInfo
        self.InterfaceProductInfo = InterfaceProductInfo
        self.ProtocolInfo = ProtocolInfo
        self.MacAddress = MacAddress
        self.OneTimePassword = OneTimePassword
        self.ClientIPAddress = ClientIPAddress
        self.AuthCode = AuthCode


RspInfo = CSecurityFtdcRspInfoField
RspUserLogin = CSecurityFtdcRspInfoField
UserLogout = CSecurityFtdcUserLogoutField
SpecificInstrument = CSecurityFtdcSpecificInstrumentField
DepthMarketData = CSecurityFtdcDepthMarketDataField
ReqUserLogin = CSecurityFtdcReqUserLoginField

File: ltsapi/api_struct.cpp
#include "api_struct.h"

static PyObject *mod = NULL;

PyObject *register_struct(PyObject * self, PyObject * args) {
    mod = PyTuple_GetItem(args, 0);
    Py_INCREF(Py_None);
    return Py_None;
}



PyObject *new_CSecurityFtdcRspInfoField(CSecurityFtdcRspInfoField * p) {

    if (p == NULL) {
        Py_INCREF(Py_None);
        return Py_None;

    }

    return PyObject_CallMethod(mod, "CSecurityFtdcRspInfoField", "is",
                               p->ErrorID, p->ErrorMsg);

}

CSecurityFtdcRspInfoField *from_CSecurityFtdcRspInfoField(PyObject * p) {
    CSecurityFtdcRspInfoField *t = (CSecurityFtdcRspInfoField *) calloc(1, sizeof(CSecurityFtdcRspInfoField));
    memset(t, 0, sizeof(CSecurityFtdcRspInfoField));
    t->ErrorID = PyInt_AsLong(PyObject_GetAttrString(p, "ErrorID"));
    strcpy(t->ErrorMsg, PyString_AsString(PyObject_GetAttrString(p, "ErrorMsg")));

    return t;
}


PyObject *new_CSecurityFtdcRspUserLoginField(CSecurityFtdcRspUserLoginField * p) {
    if (p == NULL) {
        Py_INCREF(Py_None);
        return Py_None;

    }
    return PyObject_CallMethod(mod, "CSecurityFtdcRspUserLoginField", "sssssiis",
                               p->TradingDay, p->LoginTime, p->BrokerID, p->UserID, p->SystemName, p->FrontID,
                               p->SessionID, p->MaxOrderRef);
}

CSecurityFtdcRspUserLoginField *from_CSecurityFtdcRspUserLoginField(PyObject * p) {
    CSecurityFtdcRspUserLoginField *t = (CSecurityFtdcRspUserLoginField *) calloc(1,
                                                                                  sizeof(CSecurityFtdcRspUserLoginField));
    memset(t, 0, sizeof(CSecurityFtdcRspUserLoginField));
    strcpy(t->TradingDay, PyString_AsString(PyObject_GetAttrString(p, "TradingDay")));
    strcpy(t->LoginTime, PyString_AsString(PyObject_GetAttrString(p, "LoginTime")));
    strcpy(t->BrokerID, PyString_AsString(PyObject_GetAttrString(p, "BrokerID")));
    strcpy(t->UserID, PyString_AsString(PyObject_GetAttrString(p, "UserID")));
    strcpy(t->SystemName, PyString_AsString(PyObject_GetAttrString(p, "SystemName")));
    t->FrontID = PyInt_AsLong(PyObject_GetAttrString(p, "FrontID"));
    t->SessionID = PyInt_AsLong(PyObject_GetAttrString(p, "SessionID"));
    strcpy(t->MaxOrderRef, PyString_AsString(PyObject_GetAttrString(p, "MaxOrderRef")));

    return t;
}

PyObject *new_CSecurityFtdcUserLogoutField(CSecurityFtdcUserLogoutField * p) {
    if (p == NULL) {
        Py_INCREF(Py_None);
        return Py_None;

    }
    return PyObject_CallMethod(mod, "CSecurityFtdcUserLogoutField", "ss",
                               p->BrokerID, p->UserID);
}

CSecurityFtdcUserLogoutField *from_CSecurityFtdcUserLogoutField(PyObject * p) {
    CSecurityFtdcUserLogoutField *t = (CSecurityFtdcUserLogoutField *) calloc(1, sizeof(CSecurityFtdcUserLogoutField));
    memset(t, 0, sizeof(CSecurityFtdcUserLogoutField));
    strcpy(t->BrokerID, PyString_AsString(PyObject_GetAttrString(p, "BrokerID")));
    strcpy(t->UserID, PyString_AsString(PyObject_GetAttrString(p, "UserID")));

    return t;
}



PyObject *new_CSecurityFtdcSpecificInstrumentField(CSecurityFtdcSpecificInstrumentField * p) {
    if (p == NULL) {
        Py_INCREF(Py_None);
        return Py_None;

    }
    return PyObject_CallMethod(mod, "CSecurityFtdcSpecificInstrumentField", "ss",
                               p->InstrumentID, p->ExchangeID);
}

CSecurityFtdcSpecificInstrumentField *from_CSecurityFtdcSpecificInstrumentField(PyObject * p) {
    CSecurityFtdcSpecificInstrumentField *t = (CSecurityFtdcSpecificInstrumentField *) calloc(1,
                                                                                              sizeof(CSecurityFtdcSpecificInstrumentField));
    memset(t, 0, sizeof(CSecurityFtdcSpecificInstrumentField));
    strcpy(t->InstrumentID, PyString_AsString(PyObject_GetAttrString(p, "InstrumentID")));
    strcpy(t->ExchangeID, PyString_AsString(PyObject_GetAttrString(p, "ExchangeID")));

    return t;
}


PyObject *new_CSecurityFtdcDepthMarketDataField(CSecurityFtdcDepthMarketDataField * p) {
    if (p == NULL) {
        Py_INCREF(Py_None);
        return Py_None;

    }
    return PyObject_CallMethod(mod, "CSecurityFtdcDepthMarketDataField",
                               "ssssdddddddiddddddddsididididididididididids",
                               p->TradingDay, p->InstrumentID, p->ExchangeID, p->ExchangeInstID, p->LastPrice,
                               p->PreSettlementPrice, p->PreClosePrice, p->PreOpenInterest, p->OpenPrice,
                               p->HighestPrice, p->LowestPrice, p->Volume, p->Turnover, p->OpenInterest, p->ClosePrice,
                               p->SettlementPrice, p->UpperLimitPrice, p->LowerLimitPrice, p->PreDelta, p->CurrDelta,
                               p->UpdateTime, p->UpdateMillisec, p->BidPrice1, p->BidVolume1, p->AskPrice1,
                               p->AskVolume1, p->BidPrice2, p->BidVolume2, p->AskPrice2, p->AskVolume2, p->BidPrice3,
                               p->BidVolume3, p->AskPrice3, p->AskVolume3, p->BidPrice4, p->BidVolume4, p->AskPrice4,
                               p->AskVolume4, p->BidPrice5, p->BidVolume5, p->AskPrice5, p->AskVolume5, p->AveragePrice,
                               p->ActionDay);
}

CSecurityFtdcDepthMarketDataField *from_CSecurityFtdcDepthMarketDataField(PyObject * p) {
    CSecurityFtdcDepthMarketDataField *t = (CSecurityFtdcDepthMarketDataField *) calloc(1,
                                                                                        sizeof(CSecurityFtdcDepthMarketDataField));
    memset(t, 0, sizeof(CSecurityFtdcDepthMarketDataField));
    strcpy(t->TradingDay, PyString_AsString(PyObject_GetAttrString(p, "TradingDay")));
    strcpy(t->InstrumentID, PyString_AsString(PyObject_GetAttrString(p, "InstrumentID")));
    strcpy(t->ExchangeID, PyString_AsString(PyObject_GetAttrString(p, "ExchangeID")));
    strcpy(t->ExchangeInstID, PyString_AsString(PyObject_GetAttrString(p, "ExchangeInstID")));
    t->LastPrice = PyFloat_AsDouble(PyObject_GetAttrString(p, "LastPrice"));
    t->PreSettlementPrice = PyFloat_AsDouble(PyObject_GetAttrString(p, "PreSettlementPrice"));
    t->PreClosePrice = PyFloat_AsDouble(PyObject_GetAttrString(p, "PreClosePrice"));
    t->PreOpenInterest = PyFloat_AsDouble(PyObject_GetAttrString(p, "PreOpenInterest"));
    t->OpenPrice = PyFloat_AsDouble(PyObject_GetAttrString(p, "OpenPrice"));
    t->HighestPrice = PyFloat_AsDouble(PyObject_GetAttrString(p, "HighestPrice"));
    t->LowestPrice = PyFloat_AsDouble(PyObject_GetAttrString(p, "LowestPrice"));
    t->Volume = PyInt_AsLong(PyObject_GetAttrString(p, "Volume"));
    t->Turnover = PyFloat_AsDouble(PyObject_GetAttrString(p, "Turnover"));
    t->OpenInterest = PyFloat_AsDouble(PyObject_GetAttrString(p, "OpenInterest"));
    t->ClosePrice = PyFloat_AsDouble(PyObject_GetAttrString(p, "ClosePrice"));
    t->SettlementPrice = PyFloat_AsDouble(PyObject_GetAttrString(p, "SettlementPrice"));
    t->UpperLimitPrice = PyFloat_AsDouble(PyObject_GetAttrString(p, "UpperLimitPrice"));
    t->LowerLimitPrice = PyFloat_AsDouble(PyObject_GetAttrString(p, "LowerLimitPrice"));
    t->PreDelta = PyFloat_AsDouble(PyObject_GetAttrString(p, "PreDelta"));
    t->CurrDelta = PyFloat_AsDouble(PyObject_GetAttrString(p, "CurrDelta"));
    strcpy(t->UpdateTime, PyString_AsString(PyObject_GetAttrString(p, "UpdateTime")));
    t->UpdateMillisec = PyInt_AsLong(PyObject_GetAttrString(p, "UpdateMillisec"));
    t->BidPrice1 = PyFloat_AsDouble(PyObject_GetAttrString(p, "BidPrice1"));
    t->BidVolume1 = PyInt_AsLong(PyObject_GetAttrString(p, "BidVolume1"));
    t->AskPrice1 = PyFloat_AsDouble(PyObject_GetAttrString(p, "AskPrice1"));
    t->AskVolume1 = PyInt_AsLong(PyObject_GetAttrString(p, "AskVolume1"));
    t->BidPrice2 = PyFloat_AsDouble(PyObject_GetAttrString(p, "BidPrice2"));
    t->BidVolume2 = PyInt_AsLong(PyObject_GetAttrString(p, "BidVolume2"));
    t->AskPrice2 = PyFloat_AsDouble(PyObject_GetAttrString(p, "AskPrice2"));
    t->AskVolume2 = PyInt_AsLong(PyObject_GetAttrString(p, "AskVolume2"));
    t->BidPrice3 = PyFloat_AsDouble(PyObject_GetAttrString(p, "BidPrice3"));
    t->BidVolume3 = PyInt_AsLong(PyObject_GetAttrString(p, "BidVolume3"));
    t->AskPrice3 = PyFloat_AsDouble(PyObject_GetAttrString(p, "AskPrice3"));
    t->AskVolume3 = PyInt_AsLong(PyObject_GetAttrString(p, "AskVolume3"));
    t->BidPrice4 = PyFloat_AsDouble(PyObject_GetAttrString(p, "BidPrice4"));
    t->BidVolume4 = PyInt_AsLong(PyObject_GetAttrString(p, "BidVolume4"));
    t->AskPrice4 = PyFloat_AsDouble(PyObject_GetAttrString(p, "AskPrice4"));
    t->AskVolume4 = PyInt_AsLong(PyObject_GetAttrString(p, "AskVolume4"));
    t->BidPrice5 = PyFloat_AsDouble(PyObject_GetAttrString(p, "BidPrice5"));
    t->BidVolume5 = PyInt_AsLong(PyObject_GetAttrString(p, "BidVolume5"));
    t->AskPrice5 = PyFloat_AsDouble(PyObject_GetAttrString(p, "AskPrice5"));
    t->AskVolume5 = PyInt_AsLong(PyObject_GetAttrString(p, "AskVolume5"));
    t->AveragePrice = PyFloat_AsDouble(PyObject_GetAttrString(p, "AveragePrice"));
    strcpy(t->ActionDay, PyString_AsString(PyObject_GetAttrString(p, "ActionDay")));

    return t;
}


PyObject *new_CSecurityFtdcReqUserLoginField(CSecurityFtdcReqUserLoginField * p) {
    if (p == NULL) {
        Py_INCREF(Py_None);
        return Py_None;

    }
    return PyObject_CallMethod(mod, "CSecurityFtdcReqUserLoginField", "sssssssssss",
                               p->TradingDay, p->BrokerID, p->UserID, p->Password, p->UserProductInfo,
                               p->InterfaceProductInfo, p->ProtocolInfo, p->MacAddress, p->OneTimePassword,
                               p->ClientIPAddress, p->AuthCode);
}

CSecurityFtdcReqUserLoginField *from_CSecurityFtdcReqUserLoginField(PyObject * p) {
    CSecurityFtdcReqUserLoginField *t = (CSecurityFtdcReqUserLoginField *) calloc(1,
                                                                                  sizeof(CSecurityFtdcReqUserLoginField));
    memset(t, 0, sizeof(CSecurityFtdcReqUserLoginField));
    strcpy(t->TradingDay, PyString_AsString(PyObject_GetAttrString(p, "TradingDay")));
    strcpy(t->BrokerID, PyString_AsString(PyObject_GetAttrString(p, "BrokerID")));
    strcpy(t->UserID, PyString_AsString(PyObject_GetAttrString(p, "UserID")));
    strcpy(t->Password, PyString_AsString(PyObject_GetAttrString(p, "Password")));
    strcpy(t->UserProductInfo, PyString_AsString(PyObject_GetAttrString(p, "UserProductInfo")));
    strcpy(t->InterfaceProductInfo, PyString_AsString(PyObject_GetAttrString(p, "InterfaceProductInfo")));
    strcpy(t->ProtocolInfo, PyString_AsString(PyObject_GetAttrString(p, "ProtocolInfo")));
    strcpy(t->MacAddress, PyString_AsString(PyObject_GetAttrString(p, "MacAddress")));
    strcpy(t->OneTimePassword, PyString_AsString(PyObject_GetAttrString(p, "OneTimePassword")));
    strcpy(t->ClientIPAddress, PyString_AsString(PyObject_GetAttrString(p, "ClientIPAddress")));
    strcpy(t->AuthCode, PyString_AsString(PyObject_GetAttrString(p, "AuthCode")));

    return t;
}
